### PR TITLE
API for tracking user permissions and privileges

### PR DESCRIPTION
**Pull request by stefwalter**
This adds an API called ```cockpit.permission()``` to track user permission and privileges.

The end goal of this API is to allow the UI to react to the user's privileges, by disabling buttons, etc. The end goal is not to make access control decisions on the javascript side.

### CODE REVIEW

**Comment by stefwalter** [comment]
Depends on #1643 

Further work that's necessary:
 * Documentation
 * Tests
 * Basic UI review since this changes behavior
 * Porting everything else away from check_admin()
 * Remove check_admin()
 * Add ```a[disabled] { pointer-events: none; }``` to CSS

**Comment by stefwalter** [comment]
http://stackoverflow.com/questions/10276133/how-to-disable-html-links

**Comment by stefwalter** [comment]
Installed actions can be found here: ```/usr/share/polkit-1/actions/```

**Comment by stefwalter** [comment]
If ```cockpit.permission()``` is called with a string it should be treated as a polkit action. If called with an object, then it's a set of options:
{
   "action": "org.polkit.action",
   "group": "wheel",
}

If called without any arguments, then it just checks if root.

**Comment by petervo** [comment]
I've started this work at https://github.com/petervo/cockpit/tree/permission-lookup. It's a little weird because I can't add commits to this original PR.

There is still a bit of work to do here and this commit will probably be squashed out of existence, but i figured I'd ask for a early review of what is there so far particularly the way i'm going about testing.

**Comment by stefwalter** [comment]
Could you make another pull request? We can tie the two together when merging.

**Comment by stefwalter** [comment]
Merged work from @petervo, and nice testing :)

Removed the UI changes as discussed in elsewhere. That will be a separate bug. This pull request now ends at porting check_admin to the new API. Further work is necessary once the UI is decided on.

**Comment by petervo** [comment]
Here is the commit throttling subscription callbacks.

https://github.com/petervo/cockpit/commit/0222ce7cecf77bd21c66c08c858c4f78665c3ee9

**Comment by petervo** [comment]
I added some polkit actions to other permission checks here: https://github.com/petervo/cockpit/commit/c1c5ec69a4259ee0915d1d4a91db8391d29b5cbe

We probably want to hold off on merging this until the UI decisions have been made as some integration tests fail sporadically due to  permission.allowed not being evaluated yet at the time of the check. I'm mainly showing it here because it can be used to show unprivileged user issue i was talking about earlier.

**Comment by stefwalter** [comment]
I'll merge those commits into this branch. Yes, we can wait on the UI. But I think the UI side is nearly done.

**Comment by petervo** [comment]
Copying over previous comment:

I might be missing something but In testing this I've found that making CheckAuthorization calls without the AllowUserInteraction flag returns also return False, True for unprivileged users. with the flag I get False, False. Although according to the docs we aren't supposed to pass that unless the call is being made in response to a user action. For now this doesn't result in a blocked call although that could change with #1596.

**Comment by stefwalter** [comment]
Rebased, cherry-picked patches from @petervo.

**Comment by stefwalter** [comment]
> I might be missing something but In testing this I've found that making CheckAuthorization calls without the AllowUserInteraction flag returns also return False, True for unprivileged users. with the flag I get False, False. Although according to the docs we aren't supposed to pass that unless the call is being made in response to a user action. For now this doesn't result in a blocked call although that could change with #1596.

Hmmm, very good catch. This is due to polkit wanting to offer to switch to a different identity to do the work. Example of polkit check for an non-wheel user:

```
[scruffy@falcon ~]$ pkcheck --action-id 'org.freedesktop.accounts.user-administration' --enable-internal-agent --process $$ -u
==== AUTHENTICATING FOR org.freedesktop.accounts.user-administration ===
Authentication is required to change user data
Multiple identities can be used for authentication:
 1.  Stef (stef)
 2.  Administrator (admin)
```

I'll do more research on this.

**Comment by stefwalter** [comment]
As discussed on IRC:

 * It looks like we cannot get reliable responses from polkit without setting the allow-interaction flag when calling CheckAuthorization
 * We can only pass the allow-interaction when we are sure that we're using our own cockpit-bridge agent.
 * We should streamline the effectiveness of our reauthorize polkit agent if we're going to do this all over the place, and likely cache information.

So in light of all that, I've stripped out the polkit support from this API for now. Lets migrate to the new API, then migrate to polkit checks. This allows us to proceed with other work depending on this.

All the nice mock dbus and polkit work that we did is pushed to ```stefwalter/permission-polkit``` branch, with this commit as the HEAD 81d5d4216548770c139c37f410aabf75a2ffa9e2

**Comment by stefwalter** [comment]
Ready for review.

**Comment by stefwalter** [comment]
@petervo said on IRC that this no longer has a dependency on changing the UI, as far as we can tell. That can be done separately in #1702

**Comment by stefwalter** [comment]
Fixed up documentation as well.

**Comment by petervo** [comment]
Looks good, Let me know if you want me to push to master

**Comment by petervo** [comment]
I spoke too soon, turns out there was a regression here. I pushed a fix here:

https://github.com/petervo/cockpit/commit/0c9d81142703062bff390e27050643706f7b9845

**Comment by stefwalter** [comment]
Merged that commit.

FWIW, In general we should default to ```allowed = null``` if we don't know, or if loading something failed, or still loading permission info. And in the UI disabling code, we shouldn't disable if ```allowed == null```. We should just let the user perform the task, and fail late.

File: doc/guide/api-cockpit.xml [hunk 1]
```diff
@@ -1456,6 +1456,62 @@
...
+    <para>Create a new permission object. If specified the <code>options</code> argument
+      object with a <code>"group"</code> group field. If no argument is given, the permission
+      will just check if the user is root.</para>
+
+      <variablelist>
+        <varlistentry>
+          <term><code>"allowed"</code></term>
+          <listitem><para>A Boolean that indicates whether or not the current user satisfies the given criteria. If any of the given tests evaluate to true, result will be true even if other criteria are false. Because getting the correct value may be a async operation, this value should generally be checked from within a <code>$(permission).on("changed", </code>callback.</para></listitem>
+      </varlistentry>
+    </variablelist>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: This duplicates the section below and should probably be removed.

File: doc/guide/api-cockpit.xml [hunk 2]
```diff
@@ -1456,6 +1456,62 @@
 
   </refsection>
 
+<refsection id="latest-permission">
+    <title>Permission Lookup</title>
+
+    <para>Cockpit provides a mechnisim for checking if the current user satisfies a
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: "mechanism"

File: doc/guide/api-cockpit.xml [hunk 3]
```diff
@@ -1456,6 +1456,62 @@
...
+
+    <refsection id="latest-permission-constructor">
+      <title>cockpit.permission()</title>
+<programlisting>
+ permission = cockpit.permission([options])
+</programlisting>
+
+    <para>Create a new permission object. If specified the <code>options</code> argument
+      object with a <code>"group"</code> group field. If no argument is given, the permission
+      will just check if the user is root.</para>
```
Comment (mvollmer) [inline doc/guide/api-cockpit.xml]: This doesn't make it super clear that root is always given permission.  What about
```
The "root" user is always given permission.  The <code>options</code> argument can contain a <code>group</name> field, and members of that group are also given permission.
```

File: pkg/base/cockpit.js [hunk 4]
```diff
@@ -2510,6 +2510,62 @@ function full_scope(cockpit, $) {
...
+                allowed = true;
+            else if (!group)
+                allowed = false;
+
+            if (!allowed && group && user.groups) {
+                allowed = false;
+                $.each(user.groups, function(i, name) {
+                    if (name == group) {
+                        allowed = true;
+                        return false;
```
Comment (mvollmer) [inline pkg/base/cockpit.js]: What is the significance of the return value here?

File: pkg/base/cockpit.js [hunk 5]
```diff
@@ -2510,6 +2510,62 @@ function full_scope(cockpit, $) {
...
+        self.allowed = null;
+
+        var user = cockpit.user;
+        var group = null;
+
+        if (options)
+            group = options.group;
+
+        function decide() {
+            var allowed = null;
```
Comment (mvollmer) [inline pkg/base/cockpit.js]: The logic in this function is hard to follow and hard to extend later.  Would the following be better?

```
function decide() {
  if (user.id === undefined)
     return null;

  if (user.id === 0)
     return true;

  if (group && user.groups) {
     ...
     if (name == group)
       return true;
  }

  return false;
}

function user_changed() {
  var allowed = decide();
  if (self.allowed !== allowed) {
    self.allowed = allowed;
    $(self).triggerHandler("changed");
  }
}
```

Comment (petervo) [inline pkg/base/cockpit.js]: null and false have different meanings for return values from this function so the logic here wouldn't be correct. But if your point is to return early as flow control then we can make that change.

File: pkg/base/cockpit.js (hunk 4 above)
Comment (stefwalter) [inline pkg/base/cockpit.js]: When you ```return false``` from an each callback, the iteration is cancelled. http://api.jquery.com/jquery.each/

File: doc/guide/api-cockpit.xml (hunk 1 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Makes sense.

File: doc/guide/api-cockpit.xml (hunk 3 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Makes sense.

File: doc/guide/api-cockpit.xml (hunk 2 above)
Comment (stefwalter) [inline doc/guide/api-cockpit.xml]: Roger.

**Comment by stefwalter** [comment]
Fixed review comments.

File: pkg/base/cockpit.js (hunk 4 above)
Comment (mvollmer) [inline pkg/base/cockpit.js]: Ouch, true.
